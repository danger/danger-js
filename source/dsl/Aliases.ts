// Please don't have includes in here that aren't inside the DSL folder, or the d.ts/flow defs break

export type MarkdownString = string
