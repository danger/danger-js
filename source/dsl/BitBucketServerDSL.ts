// Please don't have includes in here that aren't inside the DSL folder, or the d.ts/flow defs break

/** Key details about a repo */
export interface RepoMetaData {
  /** A path like "artsy/eigen" */
  repoSlug: string
  /** The ID for the pull/merge request "11" */
  pullRequestID: string
}

// This is `danger.bitbucket_server` inside the JSON

export interface BitBucketServerJSONDSL {
  /** The pull request and repository metadata */
  metadata: RepoMetaData
  /** The related JIRA issues */
  issues: JIRAIssue[]
  /** The PR metadata */
  pr: BitBucketServerPRDSL
  /** The commits associated with the pull request */
  commits: BitBucketServerCommit[]
  /** The comments on the pull request */
  comments: BitBucketServerPRActivity[]
  /** The activities such as OPENING, CLOSING, MERGING or UPDATING a pull request */
  activities: BitBucketServerPRActivity[]
}

export interface BitBucketServerAPIDSL {
  /** Gets the contents of a file from a repo (defaults to yours) */
  getFileContents(filePath: string, repoSlug?: string, refspec?: string): Promise<string>

  /** Make a get call against the bitbucket server API */
  get(path: string, headers: any, suppressErrors?: boolean): Promise<any>

  /** Make a post call against the bitbucket server API */
  post(path: string, headers: any, body: any, suppressErrors?: boolean): Promise<any>

  /** Make a put call against the bitbucket server API */
  put(path: string, headers: any, body: any): Promise<any>

  /** Make a delete call against the bitbucket server API */
  delete(path: string, headers: any, body: any): Promise<any>
}

// This is `danger.bitbucket_server`

/** The BitBucketServer metadata for your PR */
export interface BitBucketServerDSL extends BitBucketServerJSONDSL {
  /**
   * An authenticated API so you can extend danger's behavior.
   */
  api: BitBucketServerAPIDSL
}

/**
 * This is `danger.bitbucket_server.issues` It refers to the issues that are linked to the Pull Request.
 */
export interface JIRAIssue {
  /** The unique key for the issue e.g. JRA-11 */
  key: string
  /** The user-facing URL for that issue */
  url: string
}

// This is `danger.bitbucket_server.pr`
//
//  References:
//  -  https://developer.atlassian.com/bitbucket/api/2/reference/resource/repositories/%7Busername%7D/%7Brepo_slug%7D/pullrequests/%7Bpull_request_id%7D
//  - https://docs.atlassian.com/bitbucket-server/javadoc/4.3.2/api/reference/classes.html

/**
 * An exact copy of the PR's reference JSON. This interface has type'd the majority
 * of it for tooling's sake, but any extra metadata which BitBucket Server send
 * will still be inside the JS object.
 */

export interface BitBucketServerPRDSL {
  /** The PR's ID */
  id: number
  /** The API version */
  version: number
  /** Title of the pull request. */
  title: string
  /** The text describing the PR */
  description: string
  /** The pull request's current status. */
  state: "OPEN" | "MERGED" | "DECLINED" | "SUPERSEDED"
  /** Is the PR open? */
  open: boolean
  /** Is the PR closed? */
  closed: boolean
  /** Date PR created as number of milliseconds since the unix epoch */
  createdDate: number
  /** Date PR updated as number of milliseconds since the unix epoch */
  updatedDate: number
  /** The PR submitter's reference */
  fromRef: BitBucketServerMergeRef
  /** The repo Danger is running on */
  toRef: BitBucketServerMergeRef
  /** Was this PR locked? */
  locked: boolean
  /** The creator of the PR */
  author: BitBucketServerPRParticipant & { role: "AUTHOR" }
  /** People requested as reviewers */
  reviewers: (BitBucketServerPRParticipant & { role: "REVIEWER" })[]
  /** People who have participated in the PR */
  participants: (BitBucketServerPRParticipant & { role: "PARTICIPANT" })[]
  /** Misc links for hypermedia conformance */
  links: BitBucketServerLinks<"self">
}

// These are the individual subtypes of objects inside the larger DSL objects above.

/** A BitBucketServer specific implementation of a git commit. */
export interface BitBucketServerCommit {
  /** The SHA for the commit */
  id: string
  /** The shortened SHA for the commit */
  displayId: string
  /** The author of the commit, assumed to be the person who wrote the code. */
  author: {
    /** The id of the commit author */
    name: string
    /** The display name of the commit author */
    displayName: string
    /** The email of the commit author */
    emailAddress: string
  }
  /** The UNIX timestamp for when the commit was authored */
  authorTimestamp: number
  /** The author of the commit, assumed to be the person who commited/merged the code into a project. */
  committer: {
    /** The id of the commit committer */
    name: string
    /** The display name of the commit committer */
    displayName: string
    /** The email of the commit committer */
    emailAddress: string
  }
  /** When the commit was commited to the project */
  committerTimestamp: number
  /** The commit's message */
  message: string
  /** The commit's parents */
  parents: {
    /** The full SHA */
    id: string
    /** The simplified sha */
    displayId: string
  }[]
}

export interface BitBucketServerDiff {
  /** The file refrence when moved */
  destination?: BitBucketServerFile
  /** The original file refrence */
  source?: BitBucketServerFile
  /** A set of diff changes */
  hunks: BitBucketServerHunk[]
  /** If the hunk is massive, then it will be truncated */
  truncated: boolean
  /** The commit SHA which changed this hunk */
  toHash: string
  /** Last SHA where this hunk was changed */
  fromHash: string
  /** The settings for the whitespace */
  whitespace: "SHOW" | "IGNORE_ALL"
}

export interface BitBucketServerFile {
  components: string[]
  name: string
  parent: string
  toString: string
}

export interface BitBucketServerHunk {
  destinationLine: number
  destinationSpan: number
  segments: BitBucketServerSegment[]
  sourceLine: number
  sourceSpan: number
  truncated: boolean
}

export interface BitBucketServerSegment {
  lines: BitBucketServerLine[]
  truncated: boolean
  type: "ADDED" | "REMOVED" | "CONTEXT"
}

export interface BitBucketServerLine {
  source: number
  destination: number
  line: string
  truncated: boolean
  conflictMarker?: "OURS"
  commentIds?: number[]
}

export interface BitBucketServerPRParticipant {
  /*The user for  */
  user: BitBucketServerUser
  /** How did they contribute */
  role: "AUTHOR" | "REVIEWER" | "PARTICIPANT"
  /** Did they approve of the PR? */
  approved: boolean
  /** Their review feedback */
  status: "APPROVED" | "UNAPPROVED" | "NEEDS_WORK"
}

/**
 * A BitBucketServer user account.
 */
export interface BitBucketServerUser {
  /** The name of the user */
  name: string
  /** The email for the user */
  emailAddress: string
  /** The unique user ID */
  id: number
  /** The name to use when referencing the user */
  displayName: string
  /** Is the account active */
  active: boolean
  /** The user's slug for URLs */
  slug: string
  /** The type of a user, "NORMAL" being a typical user3 */
  type: "NORMAL" | "SERVICE"
}

/**
 * A BitBucket Server Repo
 */
export interface BitBucketServerRepo {
  /** The slug for the repo */
  slug: string
  /** The repo name */
  name?: string
  /** The type of SCM tool, probably "git" */
  scmId: string
  /** Is the repo public? */
  public: boolean
  /** Can someone fork this repo? */
  forkable: boolean
  /** Links for the projects */
  links: BitBucketServerLinks<"self" | "clone">
  /** An abstraction for grouping repos */
  project: {
    /** The project unique id */
    id: number
    /** The project's human readable project key */
    key: string
    /** Is the project publicly available */
    public: boolean
    /** The name of the project */
    name: string
    /** The project's type */
    type: string
    /** Hyperlinks for the project */
    links: BitBucketServerLinks<"self">
  }
}

export type BitBucketServerLinks<Names> = {
  [key in keyof Names]: {
    href: string
    name?: string
  }[]
}

export interface BitBucketServerMergeRef {
  id: string
  displayId: string
  latestCommit: string
  repository: BitBucketServerRepo
}

export interface BitBucketServerPRActivity {
  action: "COMMENTED" | "OPENED" | "MERGED" | "DECLINED" | "UPDATED"
  comment?: BitBucketServerPRComment
  commentAction?: "ADDED" | "UPDATED"
  commentAnchor?: {
    diffType: "COMMIT" | "EFFECTIVE" | "REQUIRED" | "RANGE"
    line: number
    lineType: "CONTEXT" | "ADDED" | "REMOVED"
    fileType: "FROM" | "TO"
    fromHash: string
    path: string
    srcPath: string
    toHash: string
  }
  createdDate: number
  id: number
  user: BitBucketServerUser
}

export interface BitBucketServerPRComment {
  author: BitBucketServerUser
  comments: BitBucketServerPRActivity[]
  createdDate: number
  updatedDate: number
  id: number
  permittedOperations: {
    deletable: boolean
    editable: boolean
  }
  text: string
  version: number
  parent?: {
    id: number
  }
}

export interface BitBucketServerPagedResponse<T> {
  size: number
  limit: number
  isLastPage: boolean
  start: number
  filter: never | null // TODO: remove never
  nextPageStart: number | null
  values: T
}

export interface BitBucketServerChangesValueAddCopyModifyDelete {
  type: "ADD" | "COPY" | "MODIFY" | "DELETE" | "UNKNOWN"
  path: {
    toString: string
  }
}

export interface BitBucketServerChangesValueMove {
  type: "MOVE"
  path: {
    toString: string
  }
  srcPath: {
    toString: string
  }
}

// prettier-ignore
export type BitBucketServerChangesValue = BitBucketServerChangesValueAddCopyModifyDelete | BitBucketServerChangesValueMove
