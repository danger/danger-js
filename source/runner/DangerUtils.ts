// The documentation for these are provided inline
// inside DangerUtilsDSL.ts

export function sentence(array: string[]): string {
  if ((array || []).length === 0) {
    return ""
  }
  if (array.length === 1) {
    return array[0]
  }
  return array.slice(0, array.length - 1).join(", ") + " and " + array.pop()
}

export function href(href?: string, text?: string): string | null {
  if (!href && !text) {
    return null
  }
  if (!href && text) {
    return text
  }
  return `<a href="${href}">${text || href}</a>`
}

export const compliment = () => {
  const compliments = ["Well done.", "Congrats.", "Woo!", "Yay.", "Jolly good show.", "Good on 'ya.", "Nice work."]
  return compliments[Math.floor(Math.random() * compliments.length)]
}
