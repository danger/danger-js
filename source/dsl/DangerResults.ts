// Please don't have includes in here that aren't inside the DSL folder, or the d.ts/flow defs break

import { Violation, isInline } from "../dsl/Violation"

/**
 * The representation of what running a Dangerfile generates.
 * This needs to be passed between processes, so data only please.
 */
export interface DangerResults {
  /**
   * Failed messages
   */
  fails: Violation[]

  /**
   * Messages for info
   */
  warnings: Violation[]

  /**
   * A set of messages to show inline
   */
  messages: Violation[]

  /**
   * Markdown messages to attach at the bottom of the comment
   */
  markdowns: Violation[]

  /** Meta information about the runtime evaluation */
  meta?: {
    /** E.g. "dangerJS", or "Danger Swift" */
    runtimeName: string
    /** e.g. "https://danger.systems/js" */
    runtimeHref: string
  }
}

export interface DangerRuntimeContainer extends DangerResults {
  /**
   * Asynchronous functions to be run after parsing
   */
  scheduled?: any[]
}

export interface DangerInlineResults {
  /**
   * Path to the file
   */
  file: string

  /**
   * Line in the file
   */
  line: number

  /**
   * Failed messages
   */
  fails: string[]

  /**
   * Messages for info
   */
  warnings: string[]

  /**
   * A set of messages to show inline
   */
  messages: string[]

  /**
   * Markdown messages to attach at the bottom of the comment
   */
  markdowns: string[]
}

/// End of Danger DSL definition

export const emptyDangerResults = {
  fails: [],
  warnings: [],
  messages: [],
  markdowns: [],
}

export function validateResults(results: DangerResults) {
  // The data we get back is JSON sent by STDIN that can come from many
  // consumers, let's take the time to ensure the data is how we think it is.
  const { fails, warnings, messages, markdowns } = results
  const props = { fails, warnings, messages, markdowns }
  Object.keys(props).forEach(name => {
    //
    // Must include the key 4 types
    if (!props[name]) {
      throw new Error(`Results passed to Danger JS did not include ${name}.\n\n${JSON.stringify(results, null, "  ")}`)
    }

    const violations: Violation[] = props[name]
    violations.forEach(v => {
      // They should always have a message
      if (!v.message) {
        throw new Error(
          `A violation passed to Danger JS in ${name} did not include \`message\`.\n\n${JSON.stringify(v, null, "  ")}`
        )
      }
      // Warn if anything other than the initial API is on a violation
      const officialAPI = ["message", "line", "file", "icon"]
      const keys = Object.keys(v).filter(f => !officialAPI.includes(f))
      if (keys.length) {
        console.warn(`Received unexpected key in Violation, expected only ${officialAPI} but got ${Object.keys(v)}`)
      }
    })
  })
}

/** Returns only the inline violations from Danger results */
export function inlineResults(results: DangerResults): DangerResults {
  return {
    fails: results.fails.filter(m => isInline(m)),
    warnings: results.warnings.filter(m => isInline(m)),
    messages: results.messages.filter(m => isInline(m)),
    markdowns: results.markdowns.filter(m => isInline(m)),
  }
}

/** Returns only the main-comment comments violations from Danger results */
export function regularResults(results: DangerResults): DangerResults {
  return {
    fails: results.fails.filter(m => !isInline(m)),
    warnings: results.warnings.filter(m => !isInline(m)),
    messages: results.messages.filter(m => !isInline(m)),
    markdowns: results.markdowns.filter(m => !isInline(m)),
    meta: results.meta,
  }
}

/** Concat all the violations into a new results */
export function mergeResults(results1: DangerResults, results2: DangerResults): DangerResults {
  return {
    fails: results1.fails.concat(results2.fails),
    warnings: results1.warnings.concat(results2.warnings),
    messages: results1.messages.concat(results2.messages),
    markdowns: results1.markdowns.concat(results2.markdowns),
    meta: results1.meta || results2.meta,
  }
}

/** Sorts all of the results according to their files and lines */
export function sortInlineResults(inlineResults: DangerInlineResults[]): DangerInlineResults[] {
  // First sort messages in every inline result
  const sortedInlineResults = inlineResults.map(i => {
    return {
      file: i.file,
      line: i.line,
      fails: i.fails.sort(),
      warnings: i.warnings.sort(),
      messages: i.messages.sort(),
      markdowns: i.markdowns.sort(),
    }
  })

  // Then sort a whole array of inline results based on file/line
  return sortedInlineResults.sort((a, b) => {
    if (a.file < b.file) {
      return -1
    } else if (a.file > b.file) {
      return 1
    } else if (a.line < b.line) {
      return -1
    } else if (a.line > b.line) {
      return 1
    } else {
      // both file & line are the same
      return 0
    }
  })
}

export function sortResults(results: DangerResults): DangerResults {
  const sortByFile = (a: Violation, b: Violation): number => {
    if (a.file === undefined) {
      return -1
    }
    if (b.file === undefined) {
      return 1
    }

    if (a.file == b.file) {
      if (a.line == undefined) {
        return -1
      }
      if (b.line == undefined) {
        return 1
      }

      if (a.line < b.line) {
        return -1
      } else if (a.line > b.line) {
        return 1
      } else {
        return 0
      }
    }

    if (a.file < b.file) {
      return -1
    } else {
      return 1
    }
  }

  return {
    fails: results.fails.sort(sortByFile),
    warnings: results.warnings.sort(sortByFile),
    messages: results.messages.sort(sortByFile),
    markdowns: results.markdowns.sort(sortByFile),
    meta: results.meta,
  }
}

export const emptyResults = (): DangerResults => ({ fails: [], markdowns: [], warnings: [], messages: [] })

export const isEmptyResults = (results: DangerResults): boolean =>
  [...results.fails, ...results.warnings, ...results.messages, ...results.markdowns].length === 0

export const isMarkdownOnlyResults = (results: DangerResults): boolean =>
  results.markdowns.length > 0 && [...results.fails, ...results.warnings, ...results.messages].length === 0

export function resultsIntoInlineResults(results: DangerResults): DangerInlineResults[] {
  // Here we iterate through all keys ("fails", "warnings", "messages", "markdowns") and for each violation
  // in given kind we produce new DangerInlineResult or append a violation to existing result. This is all
  // happening in a `violationsIntoInlineResults` function that mutates an out-of-scope variable `dangerInlineResults`.

  const dangerInlineResults: DangerInlineResults[] = []

  const violationsIntoInlineResults = (kind: string) => {
    for (let violation of results[kind]) {
      if (violation.file && violation.line) {
        const findInlineResult = dangerInlineResults.find(r => r.file == violation.file && r.line == violation.line)
        if (findInlineResult) {
          findInlineResult[kind].push(violation.message)
        } else {
          const inlineResult = {
            file: violation.file,
            line: violation.line,
            fails: [],
            warnings: [],
            messages: [],
            markdowns: [],
          }
          inlineResult[kind].push(violation.message)
          dangerInlineResults.push(inlineResult)
        }
      }
    }
  }
  Object.keys(results).forEach(violationsIntoInlineResults)

  return dangerInlineResults
}

export function inlineResultsIntoResults(inlineResults: DangerInlineResults): DangerResults {
  const messageToViolation = (message: string): Violation => {
    return { message: message, file: inlineResults.file, line: inlineResults.line }
  }

  return {
    fails: inlineResults.fails.map(messageToViolation),
    warnings: inlineResults.warnings.map(messageToViolation),
    messages: inlineResults.messages.map(messageToViolation),
    markdowns: inlineResults.markdowns.map(messageToViolation),
  }
}
