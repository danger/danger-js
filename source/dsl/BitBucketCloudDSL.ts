// Please don't have includes in here that aren't inside the DSL folder, or the d.ts/flow defs break

// TODO: extract out from BitBucket specifically, or create our own type
import { RepoMetaData } from "../dsl/BitBucketServerDSL"

export interface BitBucketCloudJSONDSL {
  /** The pull request and repository metadata */
  metadata: RepoMetaData
  /** The PR metadata */
  pr: BitBucketCloudPRDSL
  /** The commits associated with the pull request */
  commits: BitBucketCloudCommit[]
  /** The comments on the pull request */
  comments: BitBucketCloudPRComment[]
  /** The activities such as OPENING, CLOSING, MERGING or UPDATING a pull request */
  activities: BitBucketCloudPRActivity[]
}

export interface BitBucketCloudDSL extends BitBucketCloudJSONDSL {}

export interface BitBucketCloudPagedResponse<T> {
  pagelen: number
  size: number
  page: number
  next: string | undefined
  previous: string | undefined
  values: T[]
}
export interface BitBucketCloudPRDSL {
  /** The PR's ID */
  id: number
  /** Title of the pull request. */
  title: string
  /** The text describing the PR */
  description: string
  /** The pull request's current status. */
  state: "OPEN" | "MERGED" | "DECLINED" | "SUPERSEDED"
  /** Date PR created as number of milliseconds since the unix epoch */
  created_on: number
  /** Date PR updated as number of milliseconds since the unix epoch */
  updated_on: number
  /** The PR's source, The repo Danger is running on  */
  source: BitBucketCloudMergeRef
  /** The PR's destination */
  destination: BitBucketCloudMergeRef
  /** The creator of the PR */
  author: BitBucketCloudUser
  /** People requested as reviewers */
  reviewers: BitBucketCloudUser[]
  /** People who have participated in the PR */
  participants: BitBucketCloudPRParticipant[]
  /** Misc links for hypermedia conformance */
  links: BitBucketCloudLinks<
    "decline" | "commits" | "self" | "comments" | "merge" | "html" | "activity" | "diff" | "approve" | "statuses"
  >
}

export interface BitBucketCloudMergeRef {
  commit: {
    hash: string
  }
  branch: {
    name: string
  }
  repository: BitBucketCloudRepo
}

export type BitBucketCloudLinks<Names extends string> = {
  [key in Names]: {
    href: string
  }
}

export interface BitBucketCloudPRParticipant {
  /*The user who participated in this PR  */
  user: BitBucketCloudUser

  /** How did they contribute */
  role: "REVIEWER" | "PARTICIPANT"

  /** Did they approve of the PR? */
  approved: boolean
}

export interface BitBucketCloudRepo {
  name: string
  full_name: string
  uuid: string
}

export interface BitBucketCloudUser {
  /** The uuid of the commit author */
  uuid: string

  /** The display name of the commit author */
  display_name: string

  /** The nick name of the commit author */
  nickname: string

  /** The acount id of the commit author */
  account_id: string
}

/** A BitBucketCloud specific implementation of a git commit. */
export interface BitBucketCloudCommit {
  /** The SHA for the commit */
  hash: string

  /** The author of the commit, assumed to be the person who wrote the code. */
  author: {
    /** Format: `Foo Bar <foo@bar.com>` */
    raw: string
    user: BitBucketCloudUser
  }

  /** When the commit was commited to the project, in ISO 8601 format */
  date: string
  /** The commit's message */
  message: string
  /** The commit's parents */
  parents: {
    /** The full SHA */
    hash: string
  }[]

  /** The commit's links */
  links: BitBucketCloudLinks<"html">
}

export interface BitBucketCloudContent {
  raw: string
  markup: string
  html: string
  type: "rendered"
}

export interface BitBucketCloudPRComment {
  deleted: boolean
  links: BitBucketCloudLinks<"self" | "html">

  pullrequest: {
    id: number
    links: BitBucketCloudLinks<"self" | "html">
    title: string
  }
  content: BitBucketCloudContent

  /** When the comment was created, in ISO 8601 format */
  created_on: string
  user: BitBucketCloudUser

  /** When the comment was updated, in ISO 8601 format */
  updated_on: string
  type: string
  id: number

  inline?: {
    to: number | null
    from: number
    path: string
  }
}

export interface BitBucketCloudPRActivity {
  comment?: BitBucketCloudPRComment
  pull_request: {
    id: number
    title: string
  }
}
