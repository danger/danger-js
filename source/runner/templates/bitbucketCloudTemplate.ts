import { DangerResults } from "../../dsl/DangerResults"
import { Violation } from "../../dsl/Violation"
import { compliment } from "../DangerUtils"

// BitBucket Cloud supports these emojis 🎉
const noEntryEmoji = "❌"
const warningEmoji = "⚠️"
const messageEmoji = "✨"
const signatureEmoji = "🚫"
const successEmoji = "🎉"

export const dangerSignature = (results: DangerResults) => {
  let meta = results.meta || { runtimeName: "dangerJS", runtimeHref: "https://danger.systems/js" }
  return `Generated by ${signatureEmoji} [${meta.runtimeName}](${meta.runtimeHref})`
}

export const messageForResultWithIssues = `${warningEmoji}  Danger found some issues. Don't worry, everything is fixable.`

export const dangerIDToString = (id: string) => `danger-id-${id};`
export const fileLineToString = (file: string, line: number) => `  File: ${file};Line: ${line};`
/**
 * Postfix signature to be attached comment generated / updated by danger.
 */
export const dangerSignaturePostfix = (results: DangerResults, commitID?: string) => {
  let signature = dangerSignature(results)
  if (commitID !== undefined) {
    signature = `${signature} against ${commitID}`
  }
  return `

  ${signature}
  `
}

function buildMarkdownTable(header: string, defaultEmoji: string, violations: Violation[]): string {
  if (violations.length === 0 || violations.every(violation => !violation.message)) {
    return ""
  }
  return `

  |      ${violations.length} ${header} |
  | --- |
${violations.map(v => `  | ${v.icon || defaultEmoji} - ${v.message} |`).join("\n")}

  `
}

/**
 * A template function for creating a GitHub issue comment from Danger Results
 * @param {string} dangerID A string that represents a unique build
 * @param {string} commitID The hash that represents the latest commit
 * @param {DangerResults} results Data to work with
 * @returns {string} HTML
 */
export function template(dangerID: string, results: DangerResults, commitID?: string): string {
  let summaryMessage: string
  if (!results.fails.length && !results.warnings.length) {
    summaryMessage = `${successEmoji}  All green. ${compliment()}`
  } else {
    summaryMessage = messageForResultWithIssues
  }
  return `
  ${summaryMessage}

  ${buildMarkdownTable("Fails", noEntryEmoji, results.fails)}
  ${buildMarkdownTable("Warnings", warningEmoji, results.warnings)}
  ${buildMarkdownTable("Messages", messageEmoji, results.messages)}
  
  ${results.markdowns.map(v => v.message).join("\n\n")}
  
  ${dangerSignaturePostfix(results, commitID)}
  
  [](http://${dangerIDToString(dangerID)})
  `
}

export function inlineTemplate(dangerID: string, results: DangerResults, file: string, line: number): string {
  const printViolation = (defaultEmoji: string) => (violation: Violation) => {
    return `- ${violation.icon || defaultEmoji} ${violation.message}`
  }

  return `
[//]: # (${dangerIDToString(dangerID)})
[//]: # (${fileLineToString(file, line)})
${results.fails.map(printViolation(noEntryEmoji)).join("\n")}
${results.warnings.map(printViolation(warningEmoji)).join("\n")}
${results.messages.map(printViolation(messageEmoji)).join("\n")}
${results.markdowns.map(v => v.message).join("\n\n")}
  `
}
