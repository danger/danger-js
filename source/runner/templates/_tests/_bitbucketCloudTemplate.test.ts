import {
  emptyResults,
  failsResultsWithoutMessages,
  warnResults,
  failsResults,
  messagesResults,
  customIconMessagesResults,
  multipleMessagesResults,
  markdownResults,
  summaryResults,
  multipleSummaryResults,
} from "../../_tests/fixtures/ExampleDangerResults"
import { dangerSignaturePostfix, template, inlineTemplate } from "../bitbucketCloudTemplate"
import { DangerResults } from "../../../dsl/DangerResults"
import * as utils from "../../DangerUtils"

const commitID = "e70f3d6468f61a4bef68c9e6eaba9166b096e23c"

const noEntryEmoji = "❌"
const warningEmoji = "⚠️"
const messageEmoji = "✨"

describe("generating messages for BitBucket cloud", () => {
  let complimentMock: jest.Mock<string, []>
  beforeEach(() => {
    complimentMock = jest.spyOn(utils, "compliment").mockReturnValue("Well done.")
  })

  afterEach(() => {
    complimentMock.mockRestore()
  })

  it("shows no sections for empty results", () => {
    const issues = template("blankID", emptyResults, commitID)
    expect(issues).not.toContain("Fails")
    expect(issues).not.toContain("Warnings")
    expect(issues).not.toContain("Messages")
    expect(issues).toContain("Well done.")
    expect(complimentMock).toBeCalledTimes(1)
  })

  it("shows no sections for results without messages", () => {
    const issues = template("blankID", failsResultsWithoutMessages, commitID)
    expect(issues).not.toContain("Fails")
    expect(issues).not.toContain("Warnings")
    expect(issues).not.toContain("Messages")
    expect(issues).not.toContain("Well done.")
    expect(complimentMock).not.toBeCalled()
  })

  it("Shows the failing messages in a section", () => {
    const issues = template("blankID", failsResults, commitID)
    expect(issues).toContain("Fails")
    expect(issues).not.toContain("Warnings")
    expect(issues).not.toContain("Well done.")
    expect(complimentMock).not.toBeCalled()
  })

  it("Shows the warning messages in a section", () => {
    const issues = template("blankID", warnResults, commitID)
    expect(issues).toContain("Warnings")
    expect(issues).not.toContain("Fails")
    expect(issues).not.toContain("Well done.")
    expect(complimentMock).not.toBeCalled()
  })

  it("Shows custom icon for message", () => {
    const issues = template("blankID", customIconMessagesResults, commitID)
    expect(issues).toContain("Messages")
    expect(issues).toContain("📝")
    expect(issues).not.toContain(messageEmoji)
    expect(issues).not.toContain("Warnings")
    expect(issues).not.toContain("Fails")
    expect(issues).toContain("Well done.")
    expect(complimentMock).toBeCalled()
  })

  it("Mixed icon messages match snapshot", () => {
    expect(template("blankID", multipleMessagesResults, commitID)).toMatchSnapshot()
  })

  it("summary result matches snapshot, with a commit", () => {
    expect(template("blankID", summaryResults, commitID)).toMatchSnapshot()
  })

  it("summary result matches snapshot, without a commit", () => {
    expect(template("blankID", summaryResults)).toMatchSnapshot()
  })

  it("multiple summary result matches snapshot", () => {
    expect(template("blankID", multipleSummaryResults, commitID)).toMatchSnapshot()
  })

  it("shows a postfix message indicating the current commit ID at the time of comment", () => {
    expect(template("blankID", emptyResults, commitID)).toContain(dangerSignaturePostfix({} as DangerResults, commitID))
  })

  it("shows a postfix message with no commit ID if not provided", () => {
    expect(template("blankID", emptyResults)).toContain(dangerSignaturePostfix({} as DangerResults))
  })

  it("empty result matches snapshot, with a compliment", () => {
    expect(template("blankID", emptyResults, commitID)).toMatchSnapshot()
  })
})

describe("generating inline messages", () => {
  it("Shows the failing message", () => {
    const issues = inlineTemplate("blankID", failsResults, "File.swift", 5)
    expect(issues).toContain(`- ${noEntryEmoji} Failing message`)
    expect(issues).not.toContain(`- ${warningEmoji}`)
    expect(issues).not.toContain(`- ${messageEmoji}`)
  })

  it("Shows the warning message", () => {
    const issues = inlineTemplate("blankID", warnResults, "File.swift", 5)
    expect(issues).toContain(`- ${warningEmoji} Warning message`)
    expect(issues).not.toContain(`- ${noEntryEmoji}`)
    expect(issues).not.toContain(`- ${messageEmoji}`)
  })

  it("Shows the message", () => {
    const issues = inlineTemplate("blankID", messagesResults, "File.swift", 5)
    expect(issues).toContain(`- ${messageEmoji} Message`)
    expect(issues).not.toContain(`- ${noEntryEmoji}`)
    expect(issues).not.toContain(`- ${warningEmoji}`)
  })

  it("Shows message with custom icon", () => {
    const issues = inlineTemplate("blankID", customIconMessagesResults, "File.swift", 10)
    expect(issues).toContain("- 📝 Message with custom icon")
    expect(issues).not.toContain(`- ${messageEmoji}`)
    expect(issues).not.toContain(`- ${noEntryEmoji}`)
    expect(issues).not.toContain(`- ${warningEmoji}`)
  })

  it("Shows mixed messages", () => {
    const issues = inlineTemplate("blankID", multipleMessagesResults, "File.swift", 10)
    expect(issues).toContain("- 📝 Message with custom icon")
    expect(issues).toContain("- 🔔 Message with custom icon2")
    expect(issues).toContain(`- ${messageEmoji} Test message`)
    expect(issues).not.toContain(`- ${noEntryEmoji}`)
    expect(issues).not.toContain(`- ${warningEmoji}`)
  })

  it("Shows markdowns one after another", () => {
    const issues = inlineTemplate("blankID", markdownResults, "File.swift", 5)
    const expected = `
### Short Markdown Message1

### Short Markdown Message2
`
    expect(issues).toContain(expected)
  })

  it("summary inline result matches snapshot", () => {
    expect(inlineTemplate("blankID", summaryResults, "File.swift", 5)).toMatchSnapshot()
  })
})
