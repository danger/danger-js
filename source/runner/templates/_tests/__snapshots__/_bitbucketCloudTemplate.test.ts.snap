// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generating inline messages summary inline result matches snapshot 1`] = `
"
[//]: # (danger-id-blankID;)
[//]: # (  File: File.swift;Line: 5;)
- ❌ Failing message Failing message
- ⚠️ Warning message Warning message
- ✨ message
markdown
  "
`;

exports[`generating messages for BitBucket cloud Mixed icon messages match snapshot 1`] = `
"
  🎉  All green. Well done.

  
  
  

  |      3 Messages |
  | --- |
  | 📝 - Message with custom icon |
  | 🔔 - Message with custom icon2 |
  | ✨ - Test message |

  
  
  
  
  

  Generated by 🚫 [dangerJS](https://danger.systems/js) against e70f3d6468f61a4bef68c9e6eaba9166b096e23c
  
  
  [](http://danger-id-blankID;)
  "
`;

exports[`generating messages for BitBucket cloud empty result matches snapshot, with a compliment 1`] = `
"
  🎉  All green. Well done.

  
  
  
  
  
  
  

  Generated by 🚫 [dangerJS](https://danger.systems/js) against e70f3d6468f61a4bef68c9e6eaba9166b096e23c
  
  
  [](http://danger-id-blankID;)
  "
`;

exports[`generating messages for BitBucket cloud multiple summary result matches snapshot 1`] = `
"
  ⚠️  Danger found some issues. Don't worry, everything is fixable.

  

  |      2 Fails |
  | --- |
  | ❌ - Failing message Failing message |
  | ❌ - Failing message 2, This is Failing message |

  
  

  |      2 Warnings |
  | --- |
  | ⚠️ - Warning message Warning message |
  | ⚠️ - Warning message |

  
  

  |      3 Messages |
  | --- |
  | ✨ - message |
  | ✨ - message 2 |
  | ✨ - message 3 |

  
  
  markdown
  
  

  Generated by 🚫 [dangerJS](https://danger.systems/js) against e70f3d6468f61a4bef68c9e6eaba9166b096e23c
  
  
  [](http://danger-id-blankID;)
  "
`;

exports[`generating messages for BitBucket cloud summary result matches snapshot, with a commit 1`] = `
"
  ⚠️  Danger found some issues. Don't worry, everything is fixable.

  

  |      1 Fails |
  | --- |
  | ❌ - Failing message Failing message |

  
  

  |      1 Warnings |
  | --- |
  | ⚠️ - Warning message Warning message |

  
  

  |      1 Messages |
  | --- |
  | ✨ - message |

  
  
  markdown
  
  

  Generated by 🚫 [dangerJS](https://danger.systems/js) against e70f3d6468f61a4bef68c9e6eaba9166b096e23c
  
  
  [](http://danger-id-blankID;)
  "
`;

exports[`generating messages for BitBucket cloud summary result matches snapshot, without a commit 1`] = `
"
  ⚠️  Danger found some issues. Don't worry, everything is fixable.

  

  |      1 Fails |
  | --- |
  | ❌ - Failing message Failing message |

  
  

  |      1 Warnings |
  | --- |
  | ⚠️ - Warning message Warning message |

  
  

  |      1 Messages |
  | --- |
  | ✨ - message |

  
  
  markdown
  
  

  Generated by 🚫 [dangerJS](https://danger.systems/js)
  
  
  [](http://danger-id-blankID;)
  "
`;
